### ISSUE TITLE
Scaling issues with decoder cache shrinking

### ISSUE THREAD

**Issue by arvoelke** [nengo/nengo#835]
I've run into some problems with my decoder cache. `get_files()` now returns over 125,000 files every time a model is built, and this seems to be causing some problems in the `shrink()` step.

The first time I tried building a model with the newest Nengo, my Windows 8 machine blue-screened with a `SYSTEM_SERVICE_EXCEPTION`. I believe this happened on the `safe_stat` step that occurs on each of these files. The second time I tried, it worked, but took at least 2 minutes to finish this step. Now it seems to work fine each time (and is much faster), but this step still takes up the bulk of the build time.

I'm wondering if this means a smaller limit should be imposed on the decoder cache size? Also is there a way we could cache/index the shrinking step to make it faster?

**Comment by tcstewar** [comment]
Yikes..  that's worse than I've seen it on my Windows install, but every now and then I go in and wipe out that cache completely.

**Comment by tbekolay** [comment]
You can [set your own cache size limit](https://github.com/nengo/nengo/blob/master/nengo-data/nengorc), but it's in terms of file sizes, not the number of files. I can see two possible ways to alleviate this:

1. Only cache files of size greater than some threshold (1MB or something).
2. Include a configurable limit for the number of files (in addition to the overall file size limit).

**Comment by arvoelke** [comment]
(2) sounds good to me, just because (1) seems arbitrary without an understanding of how that amount maps onto the sizes of ensembles for example. Do we want to do something like benchmark the maximum number of files that can be processed by this step in X seconds? I am also curious if @jgosmann thinks some sort of indexing strategy could fix this problem without the above limits.

**Comment by jgosmann** [comment]
Might not be able to take a look at this before I'm back from my trip. But can you give me a minimal model and a recursive cache directory listing, @arvoelke?

**Comment by arvoelke** [comment]
The directory consists of 256 folders labelled `00` through `ff`. The minimal model is just a blank `nengo.Network()`, since this step happens regardless of what you're building. As to _how_ my cache may have gotten into this state, I believe it's with my "ensemble finder" experiments, where I was generating thousands of small ensembles with different seeds to find the one with the lowest RMSE. And don't worry about taking a look at this now, enjoy your trip!

**Comment by jgosmann** [comment]
@arvoelke Can you test PR #855 (branch `cache-fixes`) to see if it is any improvement?

**Comment by arvoelke** [comment]
Nope. The following script takes 5 seconds (after "warming up") on either branch:

    import nengo
    import time

    start = time.time()
    with nengo.Network() as model:
           nengo.Simulator(model)
    print time.time() - start

As stated in the original post, `get_files` is not the problem. The problem is the `safe_stat` that occurs afterwards on each of these files. Can these results be cached or moved directly into the index?

Also, I haven't looked at your code, but I was wondering if you considered how `.npz` files work internally:

> The .npz file format is a zipped archive of files named after the variables they contain... When opening the saved .npz file with load a NpzFile object is returned. 

Basically, you can just throw all of the decoders into a single `.npz` file, and then `np.load` will only read the separate file once you attempt to access its contents (see https://github.com/numpy/numpy/blob/master/numpy/lib/npyio.py#L201). Maybe this can simplify things by just placing all of the necessary metadata in the same archive alongside all of the decoders. Then you can let numpy take care of the archiving/indexing/loading/etc for you.

**Comment by jgosmann** [comment]
> The problem is the safe_stat that occurs afterwards on each of these files.

Are you sure that it is not the `os.listdir` call on a directory with a large number of files?

Also, `os.listdir` calls `stat` on each file itself. We should probably replace it with `scandir` which is included from Python 3.5 onwards. For older versions this would add [another dependency](https://github.com/benhoyt/scandir). Though it is a small module, shipping it with Nengo might be problematic as there seems to be some part implemented in C (at least for Windows).

`.npz` files are not really solving any problem. Storing the data in less files would definitely be good, but requires the implementation of our own indexing system which is the actual problem. Though, it should be doable given an efficient key-value store.

**Comment by arvoelke** [comment]
> Are you sure that it is not the os.listdir call on a directory with a large number of files?

Yes, to be sure I moved `get_files()` to a separate line, wrapped it in a timer. And then wrapped the `safe_stat` loop with a different timer:

    Time to get files: 0.0
    Time to stat files: 4.73699998856

> but requires the implementation of our own indexing system which is the actual problem.

I would have to look at the code because I'm not quite sure why this is a problem. You put whatever metadata you need (the size of the individual files you get from `os.stat`, or whatever) into the same file (as a single array, or indexed by the same key as the decoders with a '.meta' prefix), replace calls to `os.stat` with the corresponding lookup of metadata, and finally access the decoders by their key whenever you need them (in the same way you do it already -- `_get_cache_key`?). 

I'd have to take a look at the cache code before being sure about the specifics, but the point is to precompute the `safe_stat` loop, which at the same time may be simplified by taking advantage of what npz files provide for free.

**Comment by jgosmann** [comment]
Can you check that there are no error getting logged in the `safe_stat` call? (Just want to be sure about that.)

> You put whatever metadata you need (the size of the individual files you get from os.stat, or whatever) into the same file (as a single dictionary indexed by decoder key), replace calls to os.stat with the corresponding lookup of metadata

I think, I misunderstood you first (thought you wanted to merge multiple files into one). Nevertheless, to me it seems unlikely that storing the `stat` information in the file will help because instead of doing the `stat` we would end up opening every file and read from it which seems to me even more expensive (I wouldn't be surprised if `open` even makes a call to `stat` itself).

That being said, the code currently makes about three times more calls to `stat` than needed. First in the `listdir` call (with `scandir` the information from that call could be accessed), second to get the file size and third to get the file modification time.

**Comment by arvoelke** [comment]
> Can you check that there are no error getting logged in the safe_stat call? (Just want to be sure about that.)

I had checked that too. Assuming adding `logging.basicConfig(level=logging.INFO)` to an iPython notebook would work, then there were no errors getting logged.

> thought you wanted to merge multiple files into one)

Depends what you mean by "file". If you mean an `.npz` archive, then yes. But as mentioned above, internally this would be implemented by a different file per array and then lazy-loaded.  

> it seems unlikely that storing the stat information in the file will help

My suggestion was "as a single dictionary indexed by decoder key", so you would only open up one file (one `n`-by-`2` array of key-value pairs). 

 > more calls to stat than needed. First in the listdir call

Based on the previous timing information, it doesn't seem this is a real problem, though. Perhaps it has to do with the order in which the stats occur (whether their order agrees with the file system's internal index or something). Where were you getting the information that `listdir` calls `stat` on every file?

**Comment by jgosmann** [comment]
> Where were you getting the information that listdir calls stat on every file?

[Here](https://github.com/benhoyt/scandir#background), but I misread it and only `os.walk` is making the `stat` calls, but `scandir` would provide the information of `listdir` and `stat` on Windows without additional system call.

> My suggestion was "as a single dictionary indexed by decoder key", so you would only open up one file (one n-by-2 array of key-value pairs).

I have to spend some thought on how to handle concurrency in this case ... there might be multiple build processes accessing the cache in parallel and reading/writing that file at the same time.

**Comment by arvoelke** [comment]
> I have to spend some thought on how to handle concurrency in this case ... there might be multiple build processes accessing the cache in parallel and reading/writing that file at the same time.

I would be surprised if this already handled concurrency perfectly. In particular, one build might try to load a cached decoder, while another removes it to make room for more decoders. If this is actually handled, I can still imagine other examples. But regardless, I believe @tbekolay said at one point or another that we don't expect the build process to be thread-safe (although I do realize this is stricter than being process-safe). Has this policy changed since then?

**Comment by jgosmann** [comment]
I definitely ran into problems with parallelism on sharcnet when starting a huge number of parallel simulations. I think, I fixed those ... but I haven't been using the cache on sharcnet because it was slow (I should reactivate it, now that #677 is merged in).

But to me the code looks safe, at least on Unix systems (no guarantees on Windows).

**Comment by jgosmann** [comment]
@arvoelke Can you test PR #855 (branch cache-fixes) again?
You probably have to run it twice because the first run will update the cache format and will do a complete listing of the files for that. The second run should then already have that information.

Note that it is still work in progress and definitely needs more testing and some refactoring.

**Comment by jgosmann** [comment]
So, I've got a copy of Xuan's overly large cache directory to test things out. I didn't run into any problems apart from the usual uncooperativeness of Windows not related to the cache itself and a potential performance impact of the iteration over a lot of files. But with about 120000 files the cache shrink is not super fast (certainly room for improvement), but not that slow that I would consider it completely broken.

Regarding the question how did Xuan get a cache significantly larger than the 512MB limit: We don't know. My best guess is that he was using an old Nengo version (due to `git checkout -b` confusion). Other potential problems on Windows:
* Files don't get removed because they are still in use somehow.
* Building a single model produces way too many cache files and the shrink (which happens after the build) fails.

Anyways, the bottom line is: I cannot reproduce the problem, I don't know the exact cause, and I cannot decide whether any changes will help. There is PR #855 which might improve things (I will do some more performance testing with it myself), but I only want to merge it if it is a significant improvement because it increases the cache complexity quite a bit and is thus more likely to contain bugs.

**Comment by jgosmann** [comment]
PR #855 seems to be significantly slower than the current `master`. :(
I am leaning towards leaving the cache implementation as is for now.

**Comment by hunse** [comment]
Just a thought: What if we group cache results into files based on where the seed is set? For example, if the seed is set at the Network level, then all the cache entries for the ensembles in that network would be put in one file. This could really help reduce the number of files we produce, and if the seed is set at the network level, the chances are slim-to-none that the randomly generated seeds for any two ensembles will ever be the same (whether in the same network, in other networks, or other models). I realize this would require some extra code in the builder, but I wonder if it might be worth it.

**Comment by jgosmann** [comment]
I'm sceptical. I think, one of the main use cases for the cache is that you iteratively build a model and test it. In that case there will be networks with the same seed, but slightly different components. In that case the file with the grouping can either be completely rewritten which makes the cache pretty useless or it could be updated. But if it gets updated, some kind of indexing of the file contents is needed and you end up with the same problems the filesystem is already dealing with.

**Comment by jgosmann** [comment]
Note to self: I tested the performance on an SSD. I should repeat the tests on an actual hard drive to see if it makes a difference.

**Comment by hunse** [comment]
> I think, one of the main use cases for the cache is that you iteratively build a model and test it. In that case there will be networks with the same seed, but slightly different components.

Good point. We might want to limit the level at which we perform the grouping. For example, if you set the seed at the top level network, and you've got a bunch of EnsembleArrays in that network, then we could do the grouping at the level of the EnsembleArrays rather than the top level network. So generally, the idea would be that grouping happens at the lowest level network possible, rather than the highest.

Another idea is to just have EnsembleArrays be a special case, since I think that will help considerably in 99% of models, and the rest will be no worse than they are now. I still kind of wish that we made EnsembleArrays special entities, since to me they really do seem to be a special case, but that's beside the point. Given what we have now, would it be possible to do special caching for EnsembleArrays without writing a special builder for them? Also, we could consider this stuff in tandem with @arvoelke's fast EnsembleArray stuff (where is that, BTW?), since they might help each other out (though if we have fast ensemble arrays, that might fix this problem itself, since all ensembles would have the same seed and thus just need one cache entry).

**Comment by jgosmann** [comment]
Grouping for the EnsembleArrays would make sense. It might make sense to add some kind of interface so that networks can influence the caching in certain ways like grouping all decoders in one file (maybe there are other useful optimizations?). But this will probably involve some bigger changes to the builder and caching system.

Another thought: It seems there are two types of networks. Those that are pretty fixed and stay the same between builds (e.g. the networks provided by Nengo, but potentially also user defined networks) where grouping would make sense; and then those networks which are adjusted and added too between builds (the top-level network, but maybe also some user defined networks).

**Comment by jgosmann** [comment]
PR #946 hopefully fixes this issue or at least improves on the current state.

**Comment by jgosmann** [comment]
Shall we close this issue now that #946 is merged?

**Comment by arvoelke** [comment]
SGTM. Thanks!
